File: app/src/main/java/cn/sskbskdrin/record/audio/AudioTrackManager.java
package cn.sskbskdrin.record.audio;

import android.media.AudioManager;
import android.media.AudioTrack;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;

/**
 * @author sskbskdrin
 * @date 2019/April/3
 */
public class AudioTrackManager {
    private AudioTrack audioTrack;

    private boolean playing = false;

    public AudioParams playWav(String filepath, AudioRecordManager.RecordCallback callback) {
        RandomAccessFile file = null;
        AudioParams params = null;
        try {
            file = new RandomAccessFile(filepath, "r");
            params = readWavHeader(file);
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        if (params == null) {
            return null;
        }
        int simpleRate = params.simpleRate;
        int channelConfig = params.getOutChannelConfig();
        int audioFormat = params.getEncodingFormat();
        int minBufSize = AudioTrack.getMinBufferSize(simpleRate, channelConfig, audioFormat);
        audioTrack = new AudioTrack(AudioManager.STREAM_MUSIC, simpleRate, channelConfig, audioFormat, minBufSize,
            AudioTrack.MODE_STREAM);
        final RandomAccessFile finalFile = file;
        new Thread(() -> {
            playing = true;
            audioTrack.play();
            byte[] buffer = new byte[minBufSize];
            try {
                finalFile.seek(44);
                while (playing) {
                    int read = finalFile.read(buffer);
                    if (read > 0) {
                        audioTrack.write(buffer, 0, read);
                    } else {
                        finalFile.close();
                        playing = false;
                        audioTrack.stop();
                        audioTrack.release();
                    }
                    if (callback != null) {
                        callback.onRecord(buffer, read);
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }).start();
        return params;
    }

    public static AudioParams readWavHeader(RandomAccessFile file) throws IOException {
        file.seek(22);
        byte channelCount = file.readByte();
        file.seek(24);
        int sampleRate = file.readByte() & 0xff;
        sampleRate |= (file.readByte() & 0xff) << 8;
        sampleRate |= (file.readByte() & 0xff) << 16;
        sampleRate |= (file.readByte() & 0xff) << 24;

        file.seek(34);
        byte bits = file.readByte();
        return new AudioParams(sampleRate, channelCount, bits);
    }
}

File: kotlinapp/src/main/java/cn/sskbskdrin/audiovideo/audio/AudioTrackManager.kt
package cn.sskbskdrin.record.audio

import android.media.AudioManager
import android.media.AudioTrack
import java.io.RandomAccessFile
import kotlin.concurrent.thread


/**
 * @author sskbskdrin
 * @date 2019/March/23
 */

fun readWavHeader(file: RandomAccessFile): AudioParams {
    file.seek(22)
    val channelCount = file.readByte()
    file.seek(24)
    var byte = file.readByte().toInt() and 0xff
    var sampleRate = byte
    byte = file.readByte().toInt() and 0xff
    sampleRate = byte shl 8 or sampleRate
    byte = file.readByte().toInt() and 0xff
    sampleRate = byte shl 16 or sampleRate
    byte = file.readByte().toInt() and 0xff
    sampleRate = byte shl 24 or sampleRate

    file.seek(34)
    val bits = file.readByte()
    return AudioParams(sampleRate, channelCount.toInt(), bits.toInt())
}

class AudioTrackManager {
    private var audioTrack: AudioTrack? = null

    private var playing = false

    fun playWav(filepath: String, block: (data: ByteArray, length: Int) -> Unit = { _, _ ->
    }): AudioParams {
        val file = RandomAccessFile(filepath, "r")
        val params = readWavHeader(file)
        val simpleRate = params.simpleRate
        val channelConfig = params.getOutChannelConfig()
        val audioFormat = params.getEncodingFormat()
        val minBufSize = AudioTrack.getMinBufferSize(simpleRate, channelConfig, audioFormat)
        audioTrack = AudioTrack(AudioManager.STREAM_MUSIC,
                simpleRate,
                channelConfig,
                audioFormat,
                minBufSize,
                AudioTrack.MODE_STREAM)
        thread(start = true) {
            playing = true
            audioTrack?.play()
            val buffer = ByteArray(minBufSize)
            file.seek(44)
            while (playing) {
                val read = file.read(buffer)
                if (read > 0) {
                    audioTrack?.write(buffer, 0, read)
                } else {
                    file.close()
                    playing = false
                    audioTrack?.stop()
                }
                block(buffer, read)
            }
        }
        return params
    }
}